File: MoneyAI.Repositories/StatementParsers/PayPalParser.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using CommonUtils;

namespace MoneyAI.Repositories.StatementParsers
{
    internal class PayPalParser : GenericStatementParser
    {
        public PayPalParser(string filePath)
            : base(filePath, new string[] { ".csv", ".iif" })
        {
        }

        protected override StatementParserBase.StatementColumnType GetColumnType(string columnName)
        {
            return StatementColumnType.ProviderAttribute;
        }

        protected override void SetCalculatedAttributes(Transaction.ImportedValues importedValues)
        {
            base.SetCalculatedAttributes(importedValues);

            var dateString = importedValues.ProviderAttributes.GetValueOrDefault(@"date");
            var timeZoneAbbreviation = importedValues.ProviderAttributes.GetValueOrDefault(@"time zone");
            string timeZoneHoursString = null;
            if (!string.IsNullOrWhiteSpace(timeZoneAbbreviation))
                timeZoneHoursString = Utils.GetTimeZoneHoursFromAbbreviation(timeZoneAbbreviation);
            var timeString = importedValues.ProviderAttributes.GetValueOrDefault(@"time");

            var dateTimeString = string.Concat(dateString, " ", timeString, " ", timeZoneHoursString);
            importedValues.TransactionDate = DateTime.Parse(dateTimeString);    //TODO: handle local

            var memo = importedValues.ProviderAttributes.GetValueOrDefault(@"memo");
            var name = importedValues.ProviderAttributes.GetValueOrDefault(@"name");
            importedValues.EntityName =  (!string.IsNullOrWhiteSpace(memo) ? string.Concat(memo, " - ", name) : name);

            var amountString = importedValues.ProviderAttributes.GetValueOrDefault(@"amount");
            importedValues.Amount = this.ParseAmount(importedValues.ProviderAttributes[@"amount"]);

            var payPalType = importedValues.ProviderAttributes.GetValueOrDefault(@"type") ??    //CSV format
                importedValues.ProviderAttributes.GetValueOrDefault(@"class");  //IIF format

            SetTransactionreason(importedValues, payPalType);

            importedValues.ContentHash = Transaction.GetContentHash(
                Utils.AsEnumerable(name, dateString, amountString, payPalType).Select(s => s.ToUpperInvariant()));
        }

        private static void SetTransactionreason(Transaction.ImportedValues importedValues, string payPalType)
        {
            var ignorableActivity = false;
            switch (importedValues.ProviderAttributes.GetValueOrDefault(@"status"))
            {
                case "Denied":
                case "Removed":
                case "Placed":
                case "Canceled":
                case "Cleared":
                case "Failed":
                case "Refunded":
                    ignorableActivity = true;
                    break;
            }

            if (!ignorableActivity)
            {
                if (payPalType.StartsWith("Update to ", StringComparison.InvariantCultureIgnoreCase) ||
                    payPalType.StartsWith("Cancelled", StringComparison.InvariantCultureIgnoreCase) ||
                    payPalType.StartsWith("Failed", StringComparison.InvariantCultureIgnoreCase) ||
                    payPalType.StartsWith("Denied", StringComparison.InvariantCultureIgnoreCase))
                {
                    ignorableActivity = true;
                }
                else
                {
                    switch (payPalType)
                    {
                        case "Authorization":
                        case "Temporary Hold":
                            ignorableActivity = true;
                            break;
                    }
                }
            }

            if (!ignorableActivity)
            {
                if (payPalType.IndexOf("Payment Sent", StringComparison.InvariantCultureIgnoreCase) >= 0)
                    importedValues.TransactionReason = TransactionReason.Purchase;
                else if (payPalType.IndexOf("Donation Sent", StringComparison.InvariantCultureIgnoreCase) >= 0)
                    importedValues.TransactionReason = TransactionReason.Purchase;
                else if (payPalType.IndexOf("Refund", StringComparison.InvariantCultureIgnoreCase) >= 0)
                    importedValues.TransactionReason = TransactionReason.Return;
                else if (payPalType.IndexOf("Charge From ", StringComparison.InvariantCultureIgnoreCase) >= 0 &&
                    payPalType.IndexOf("Card", StringComparison.InvariantCultureIgnoreCase) >= 0)
                    importedValues.TransactionReason = TransactionReason.InterAccountPayment;
                else if (payPalType.IndexOf("Credit To ", StringComparison.InvariantCultureIgnoreCase) >= 0 &&
                    payPalType.IndexOf("Card", StringComparison.InvariantCultureIgnoreCase) >= 0)
                    importedValues.TransactionReason = TransactionReason.InterAccountTransfer;
                else if (payPalType.IndexOf("Add Funds from a Bank Account", StringComparison.InvariantCultureIgnoreCase) >= 0)
                    importedValues.TransactionReason = TransactionReason.InterAccountTransfer;
                else
                {
                    switch (payPalType)
                    {
                        case "BillPay":
                            importedValues.TransactionReason = TransactionReason.Purchase;
                            break;
                        case "Payment Received":
                            importedValues.TransactionReason = TransactionReason.PayMentRecieved;
                            break;
                        default:
                            throw new Exception("Cannot determine TransactionReason for PayPal Type '{0}'".FormatEx(payPalType));
                    }
                }
            }
            else
                importedValues.TransactionReason = null;
        }

        protected override TransactionReason? InferTransactionReason(Transaction.ImportedValues importedValues)
        {
            return importedValues.TransactionReason;    //Disable default inference
        }

        protected override bool ValidateImportedValues(Transaction.ImportedValues importedValues)
        {
            if (importedValues.TransactionReason == null)
                return false;

            return base.ValidateImportedValues(importedValues);
        }
    }
}

File: packages/server/src/parsers/statement/paypal-parser.ts
/**
 * PayPal statement parser, ported from C# PayPalParser.
 *
 * PayPal CSV/IIF files have unique column names, date+time+timezone assembly,
 * activity filtering, type-to-reason mapping, and custom content hash.
 *
 * @module
 */

import type { ImportedValues } from '@moneyinmotion/core';
import { TransactionReason, isNullOrWhitespace } from '@moneyinmotion/core';
import { getMD5HashString } from '@moneyinmotion/core';
import type { ParsedRow } from '../file-format/types.js';
import { ContentType, createFileFormatParser } from '../file-format/index.js';
import { StatementParserBase, StatementColumnType } from './statement-parser-base.js';

// ---------------------------------------------------------------------------
// Timezone abbreviation to UTC offset mapping (subset)
// ---------------------------------------------------------------------------

const timezoneOffsets: Record<string, string> = {
  PST: '-08:00',
  PDT: '-07:00',
  MST: '-07:00',
  MDT: '-06:00',
  CST: '-06:00',
  CDT: '-05:00',
  EST: '-05:00',
  EDT: '-04:00',
  GMT: '+00:00',
  UTC: '+00:00',
};

export class PayPalParser extends StatementParserBase {
  private parsedRows: ParsedRow[];

  constructor(content: string, contentType: ContentType) {
    super();

    const allowedTypes = [ContentType.Csv, ContentType.QuickBooksIif];
    if (!allowedTypes.includes(contentType)) {
      throw new Error(
        `Content type "${contentType}" is not supported for PayPal. Allowed: ${allowedTypes.join(', ')}`,
      );
    }

    const parser = createFileFormatParser(contentType);
    this.parsedRows = parser.parse(content);
  }

  protected getTransactionProperties(): ParsedRow[] {
    return this.parsedRows;
  }

  /**
   * Map all PayPal columns to ProviderAttribute so that
   * setCalculatedAttributes can do the custom mapping.
   */
  protected override getColumnType(_columnName: string): StatementColumnType {
    return StatementColumnType.ProviderAttribute;
  }

  protected override setCalculatedAttributes(importedValues: ImportedValues): void {
    super.setCalculatedAttributes(importedValues);

    const attrs = importedValues.providerAttributes ?? {};

    // Assemble date+time+timezone
    const dateString = attrs['date'] ?? '';
    const timeZoneAbbreviation = attrs['time zone'] ?? '';
    let timeZoneHoursString = '';
    if (!isNullOrWhitespace(timeZoneAbbreviation)) {
      const normalizedTimeZone = timeZoneAbbreviation.trim().toUpperCase();
      timeZoneHoursString = timezoneOffsets[normalizedTimeZone] ?? '';
      if (!timeZoneHoursString) {
        throw new Error(`Unsupported PayPal time zone: "${timeZoneAbbreviation}"`);
      }
    }
    const timeString = attrs['time'] ?? '';

    const dateTimeString = `${dateString} ${timeString} ${timeZoneHoursString}`.trim();
    const parsed = new Date(dateTimeString);
    if (isNaN(parsed.getTime())) {
      throw new Error(`Cannot parse PayPal transaction date: "${dateTimeString}"`);
    }
    importedValues.transactionDate = parsed.toISOString();

    // Entity name from memo + name
    const memo = attrs['memo'] ?? '';
    const name = attrs['name'] ?? '';
    importedValues.entityName = !isNullOrWhitespace(memo) ? `${memo} - ${name}` : name;

    // Amount
    importedValues.amount = this.parseAmount(attrs['amount'] ?? '0');

    // PayPal type from CSV "type" or IIF "class"
    const payPalType = attrs['type'] ?? attrs['class'] ?? '';

    setPayPalTransactionReason(importedValues, payPalType);

    // Custom content hash
    const amountString = attrs['amount'] ?? '';
    importedValues.contentHash = getMD5HashString(
      [name, dateString, amountString, payPalType].map((s) => (s ?? '').toUpperCase()).join('\t'),
      true,
    );
  }

  /**
   * Disable default inference -- PayPal sets reason explicitly.
   */
  protected override inferTransactionReason(importedValues: ImportedValues): number {
    return importedValues.transactionReason ?? TransactionReason.Purchase;
  }

  protected override validateImportedValues(importedValues: ImportedValues): boolean {
    if (importedValues.transactionReason == null) {
      return false;
    }
    return super.validateImportedValues(importedValues);
  }
}

// ---------------------------------------------------------------------------
// PayPal-specific transaction reason mapping
// ---------------------------------------------------------------------------

function setPayPalTransactionReason(importedValues: ImportedValues, payPalType: string): void {
  const attrs = importedValues.providerAttributes ?? {};
  let ignorableActivity = false;

  // Check status
  const status = (attrs['status'] ?? '').trim().toLowerCase();
  switch (status) {
    case 'denied':
    case 'removed':
    case 'placed':
    case 'canceled':
    case 'cancelled':
    case 'cleared':
    case 'failed':
    case 'refunded':
      ignorableActivity = true;
      break;
  }

  if (!ignorableActivity) {
    const typeLower = payPalType.toLowerCase();
    if (
      typeLower.startsWith('update to ') ||
      typeLower.startsWith('cancelled') ||
      typeLower.startsWith('failed') ||
      typeLower.startsWith('denied')
    ) {
      ignorableActivity = true;
    } else {
      switch (typeLower) {
        case 'authorization':
        case 'temporary hold':
          ignorableActivity = true;
          break;
      }
    }
  }

  if (!ignorableActivity) {
    const typeLower = payPalType.toLowerCase();
    if (typeLower.includes('payment sent')) {
      importedValues.transactionReason = TransactionReason.Purchase;
    } else if (typeLower.includes('donation sent')) {
      importedValues.transactionReason = TransactionReason.Purchase;
    } else if (typeLower.includes('refund')) {
      importedValues.transactionReason = TransactionReason.Return;
    } else if (typeLower.includes('charge from ') && typeLower.includes('card')) {
      importedValues.transactionReason = TransactionReason.InterAccountPayment;
    } else if (typeLower.includes('credit to ') && typeLower.includes('card')) {
      importedValues.transactionReason = TransactionReason.InterAccountTransfer;
    } else if (typeLower.includes('add funds from a bank account')) {
      importedValues.transactionReason = TransactionReason.InterAccountTransfer;
    } else {
      switch (typeLower) {
        case 'billpay':
          importedValues.transactionReason = TransactionReason.Purchase;
          break;
        case 'payment received':
          importedValues.transactionReason = TransactionReason.PaymentRecieved;
          break;
        default:
          throw new Error(`Cannot determine TransactionReason for PayPal Type '${payPalType}'`);
      }
    }
  } else {
    // Clear the reason so `validateImportedValues` (which checks
    // `transactionReason == null`) skips this row downstream.
    // ImportedValues.transactionReason is declared as `number` (required)
    // but the base class actually builds rows via `Partial<ImportedValues>`,
    // so removing the property at runtime is safe.
    Reflect.deleteProperty(importedValues, 'transactionReason');
  }
}
